File: algorithms/Combination Sum II/Combination Sum II.cpp
class Solution {
public:
    vector<vector<int>> ans;vector<int> temp;
    void f(vector<int>& nums,int target,int i){
        if(target==0){
            ans.push_back(temp);
            return;
        }
        if(i>=nums.size())
            return;
        if(nums[i]<=target){
            temp.push_back(nums[i]);
            f(nums,target-nums[i],i+1);
            temp.pop_back();
            while(i<nums.size()-1 && nums[i]==nums[i+1]) i++;
            f(nums,target,i+1);
        }
        else
            f(nums,target,i+1);
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(),candidates.end());
        f(candidates,target,0);
        return ans;
    }
};

File: algorithms/Combination Sum II/Combination Sum II.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function(candidates, target) {
    candidates.sort((a, b) => a - b);
    
    const ans = [];
    
    function dfs(idx, t, st) {
        if (t === 0) {
            ans.push(Array.from(st));
            return;
        }
        for (let i = idx; i < candidates.length; i++) {
            if (i > idx && candidates[i - 1] === candidates[i]) continue;
            if (candidates[i] > t) break;
            st.push(candidates[i]);
            dfs(i + 1, t - candidates[i], st);
            st.pop();
        }
    }
    
    dfs(0, target, []);
    
    return ans;
};
